src/app/clusters/scenes/SceneTable.h

```diff
@@ -0,0 +1,293 @@
+
+/**
+ * @brief Indicates the absence of a Scene table entry.
+ */
+#ifndef CHIP_CONFIG_SCENES_TABLE_NULL_INDEX
+#define CHIP_CONFIG_SCENES_TABLE_NULL_INDEX 0xFF
+#endif
+
+/**
+ * @brief The group identifier for the global scene.
+ */
+#ifndef CHIP_CONFIG_SCENES_GLOBAL_SCENE_GROUP_ID
+#define CHIP_CONFIG_SCENES_GLOBAL_SCENE_GROUP_ID 0x0000
+#endif
+
+/**
+ * @brief The scene identifier for the global scene.
+ */
+#ifndef CHIP_CONFIG_SCENES_GLOBAL_SCENE_SCENE_ID
+#define CHIP_CONFIG_SCENES_GLOBAL_SCENE_SCENE_ID 0x00
+#endif
+
+namespace chip {
+namespace scenes {
+
+typedef uint32_t TransitionTimeMs;
+typedef uint16_t SceneTransitionTime;
+typedef uint8_t TransitionTime100ms;
+
+constexpr SceneGroupID kGlobalGroupSceneId  = CHIP_CONFIG_SCENES_GLOBAL_SCENE_GROUP_ID;
+constexpr SceneIndex kUndefinedSceneIndex   = 0xff;
+constexpr SceneId kUndefinedSceneId         = CHIP_CONFIG_SCENES_TABLE_NULL_INDEX;
+static constexpr uint8_t kMaxScenePerFabric = CHIP_CONFIG_SCENES_MAX_PER_FABRIC;
+static constexpr uint8_t kMaxSceneHandlers  = CHIP_CONFIG_SCENES_MAX_CLUSTERS_PER_SCENES;
+static constexpr size_t kIteratorsMax       = CHIP_CONFIG_MAX_SCENES_CONCURRENT_ITERATORS;
+static constexpr size_t kSceneNameMax       = CHIP_CONFIG_SCENES_CLUSTER_MAXIMUM_NAME_LENGTH;
+
+/// @brief Abstract class allowing different Endpoints interactions with the ExtensionFieldSets added to and retrieved from the
+/// scene Table
+class SceneHandler
+{
+public:
+    SceneHandler(){};
+    virtual ~SceneHandler() = default;
+
+    /// @brief Copies the list of supported clusters for an endpoint in a Span and resizes the span to fit the actual number of
+    /// supported clusters
+    /// @param endpoint target endpoint
+    /// @param clusterBuffer Buffer to hold the supported cluster IDs, cannot hold more than
+    /// CHIP_CONFIG_SCENES_MAX_CLUSTERS_PER_SCENES, the function shall use the reduce_size() method in the event it is supporting
+    /// less than CHIP_CONFIG_SCENES_MAX_CLUSTERS_PER_SCENES clusters
+    virtual void GetSupportedClusters(EndpointId endpoint, Span<ClusterId> & clusterBuffer) = 0;
+
+    /// @brief Returns whether or not a cluster for scenes is supported on an endpoint
+
+    /// @param endpoint Target Endpoint ID
+    /// @param cluster Target Cluster ID
+    /// @return true if supported, false if not supported
+    virtual bool SupportsCluster(EndpointId endpoint, ClusterId cluster) = 0;
+
+    /// @brief From command AddScene, allows handler to filter through clusters in command to serialize only the supported ones.
+    /// @param endpoint Endpoint ID
+    /// @param cluster  Cluster ID to fetch from command
+    /// @param serialisedBytes Buffer for ExtensionFieldSet in command
+    /// @param extensionFieldSet ExtensionFieldSets provided by the AddScene Command
+    /// @return CHIP_NO_ERROR if successful, CHIP_ERROR value otherwise
+    virtual CHIP_ERROR SerializeAdd(EndpointId endpoint, ClusterId & cluster, MutableByteSpan & serialisedBytes,

```

Comment:
Why is `cluster` passed by reference here?  Does the callee mutate it?  If not, it should be passed by value.

Reply:
Because this function is fetching the cluster ID from the cluster-object, thus mutating it.

Reply:
I think this function could benefit from a clear description of what is input and what is output.  For example, is `extensionFieldSet` input or output?  It looks like output, since it's not `const`, but it looks like it's intended to be input.

I suggest:

1) Putting the inputs before the outputs in the function signature.
2) Clearly documenting the @params as [in] or [out].